bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,165 @@
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self.node = None
+        self._descriptor_nodes = {}  # references to descriptor containers by uid's
+        self._SR_nodes = {}
+        self._SR_cache = {}
+
+    def start(self, doc):
+        self.node = self.client.create_container(
+            key=doc["uid"],
+            metadata={"start": doc},
+            specs=[Spec("BlueskyRun", version="1.0")],
+        )
+
+    def stop(self, doc):
+        metadata = dict(self.node.metadata) | {"stop": doc}
+        self.node.update_metadata(metadata=metadata)
+
+    def descriptor(self, doc):
+        descriptor_node = self.node.create_container(key=doc["name"], metadata=doc)
+        self._descriptor_nodes[doc["uid"]] = descriptor_node
+        descriptor_node.create_container(key="external")
+        descriptor_node.create_container(key="internal")
+
+    def event(self, doc):
+        descriptor_node = self._descriptor_nodes[doc["descriptor"]]
+        parent_node = descriptor_node["internal"]
+        for table_key in ["data", "timestamps"]:
+            df = pd.DataFrame(
+                {column: [value] for column, value in doc[table_key].items()}
+            )
+            if table_key in parent_node:
+                parent_node[table_key].append_partition(df, 0)
+            else:
+                parent_node.new(
+                    structure_family=StructureFamily.table,
+                    data_sources=[
+                        DataSource(
+                            structure_family=StructureFamily.table,
+                            structure=TableStructure.from_pandas(df),
+                            mimetype="text/csv",
+                        ),  # or PARQUET_MIMETYPE
+                    ],
+                    key=table_key,
+                )
+                parent_node[table_key].write_partition(df, 0)
+
+    def stream_resource(self, doc):
+        # Only cache the StreamResource; add the node when at least one StreamDatum is added
+        self._SR_cache[doc["uid"]] = doc
+
+    def stream_datum(self, doc):
+        descriptor_node = self._descriptor_nodes[doc["descriptor"]]
+        parent_node = descriptor_node["external"]
+
+        num_rows = (
+            doc["indices"]["stop"] - doc["indices"]["start"]
+        )  # Number of rows added by new StreamDatum
+
+        # Get the Stream Resource node if it already exists or register if from a cached SR document
+        try:
```

Comment:
Elsewhere, the decision of "condition vs exception" came up. This is another place where using a condition might be cleaner.

Reply:
Done!